custom/icds_reports/reports/service_delivery_dashboard_data.py

```diff
@@ -0,0 +1,352 @@
+import copy
+from datetime import date
+
+from custom.icds_reports.cache import icds_quickcache
+from custom.icds_reports.models.views import ServiceDeliveryReportView
+from custom.icds_reports.utils import DATA_NOT_ENTERED, percent_or_not_entered, apply_exclude, \
+    format_data_not_entered_to_zero
+
+
+@icds_quickcache([
+    'domain', 'start', 'length', 'order', 'reversed_order', 'location_filters', 'year', 'month', 'step'
+], timeout=30 * 60)
+def get_service_delivery_report_data(domain, start, length, order, reversed_order, location_filters,
+                                     year, month, step, include_test=False):
+    if location_filters.get('aggregation_level') == 1:
+        default_order = 'state_name'
+    elif location_filters.get('aggregation_level') == 2:
+        default_order = 'district_name'
+    elif location_filters.get('aggregation_level') == 3:
+        default_order = 'block_name'
+    elif location_filters.get('aggregation_level') == 4:
+        default_order = 'supervisor_name'
+    else:
+        default_order = 'awc_name'
+
+    location_fields = ['state_name', 'district_name', 'block_name', 'supervisor_name', 'awc_name']
+    value_fields = location_fields + ['num_launched_awcs', 'valid_visits', 'expected_visits', 'gm_0_3',
+                                      'children_0_3', 'num_awcs_conducted_cbe', 'num_awcs_conducted_vhnd',
+                                      'thr_21_days', 'thr_25_days', 'thr_eligible', 'lunch_21_days',
+                                      'lunch_25_days', 'pse_eligible', 'pse_21_days', 'pse_25_days',
+                                      'gm_3_5', 'children_3_5']
+    data = ServiceDeliveryReportView .objects.filter(
+        month=date(year, month, 1),
+        **location_filters
+    ).order_by(default_order).values(*value_fields)
+    if not include_test:
+        data = apply_exclude(domain, data)
+    data_count = data.count()
+    config = {
+        'data': [],
+    }
+
+    def should_show_25():
+        return year >= 2020 and month >= 4
+
+    def get_value_or_data_not_entered(source, field):
+        value = source.get(field)
+        if value is None:
+            return DATA_NOT_ENTERED
+        return value
+
+    def update_all_row(first_dict, second_dict):
```

Comment:
could you rename to `update_total_row` currently it feels like it is updating all rows

Reply:
done